### ISSUE TITLE
Lock files leave dangling file descriptors

### ISSUE THREAD

**Issue by madkinsz** [mwilliamson/locket.py#14]
See https://github.com/dask/distributed/pull/6122

The `LockFile` implementation [opens a file descriptor to check locks](https://github.com/mwilliamson/locket.py/blob/master/locket/__init__.py#L138-L140). In the non-blocking case, if the lock cannot be acquired, a `LockError` is thrown and an open file descriptor is left dangling. In normal operation, this file is closed on lock release, but since the lock has already been acquired by another instance of the `LockFile`, release never occurs. On teardown, a `ResourceWarning` is thrown by the Python runtime.

I'm happy to contribute a fix here.

**Comment by mwilliamson** [comment]
Hmm, so the fix is something like this? Or have I misunderstood?

```python
    def acquire(self, timeout=None, retry_period=None):
        fileobj = open(self._path, "wb")

        try:
            if timeout is None and _lock_file_blocking_available:
                _lock_file_blocking(fileobj)
            else:
                _acquire_non_blocking(
                    acquire=lambda: _lock_file_non_blocking(fileobj),
                    timeout=timeout,
                    retry_period=retry_period,
                    path=self._path,
                )

        except:
            fileobj.close()
            raise

        else:
            self._file = fileobj
```

**Comment by madkinsz** [comment]
Yeah that'd work if you want to retain caching of the file descriptor on the object!

I opted to remove the caching and just use a context managed file in the PR to Dask, but I think they're equally reasonable. If you opt for this implementation, you'll need to add handling for a null `self._file` in `release` as well.

**Comment by madkinsz** [comment]
After thinking a little more, you may run into issues with open file descriptors if many file locks are being used. I'd suggest dropping the open descriptor from the object and just opening/closing on use unless there are explicit performance benefits to keeping it open.

**Comment by mwilliamson** [comment]
> After thinking a little more, you may run into issues with open file descriptors if many file locks are being used. I'd suggest dropping the open descriptor from the object and just opening/closing on use unless there are explicit performance benefits to keeping it open.

Hmm, I haven't looked at how the locking APIs (e.g. fcntl) work in a long time. I would have assumed that if you close the file, you lose the lock? Presumably that's not the case if the tests are all still passing with that change?

**Comment by madkinsz** [comment]
That's a great question. The tests aren't passing in Dask with my changes, but there are a lot of other side-effects to worry about there. I'll do some research and report back.

**Comment by mwilliamson** [comment]
master now has some changes that should address this, if you want to try them out?

**Comment by madkinsz** [comment]
Giving this a test at https://github.com/dask/distributed/pull/6166/commits/cab09ec919a56421e1e0ef90e72f6cd3dfd44cc5

**Comment by mwilliamson** [comment]
These changes should be available as 1.0.0 on PyPI now.
